File: Server_Root/Server/RoomServer/Behavior/Behaviors/Fire.cpp
#include "stdafx.h"
#include "Fire.h"
#include "../../Entity/Entities/Projectile/Projectile.h"
#include "../BehaviorIDs.h"
#include "../../MasterData/MasterDataManager.h"
#include "../../MasterData/Behavior.h"
#include "../../Factory.h"
#include "../../Game/BaeGameRoom.h"
#include "../../GameEvent/GameEvents/EntityCreate.h"
#include "../../Util.h"
#include "../../State/StateIDs.h"

const string Fire::NAME = "Fire";

Fire::Fire(BaeGameRoom* pGameRoom, IEntity* pEntity, int nMasterDataID) : IBehavior(pGameRoom, pEntity, nMasterDataID)
{
}

Fire::~Fire()
{
}

void Fire::Start(long long lStartTime, ...)
{
	__super::Start(lStartTime);

	va_list ap;
	va_start(ap, lStartTime);
	m_pTarget = va_arg(ap, IEntity*);
	va_end(ap);
}

void Fire::Initialize()
{
	MasterData::Behavior* pMasterBehavior = NULL;
	MasterDataManager::Instance()->GetData<MasterData::Behavior>(m_nMasterDataID, pMasterBehavior);

	m_fLength = pMasterBehavior->m_fLength;
	m_strStringParams = pMasterBehavior->m_strStringParams;
	m_vecAction = pMasterBehavior->m_vecAction;
}

void Fire::UpdateBody(long long lUpdateTime)
{
	for (vector<MasterData::Behavior::Action>::iterator it = m_vecAction.begin(); it != m_vecAction.end(); ++it)
	{
		if ((m_fCurrentTime == 0 && it->m_fTime == 0) || (m_fPreviousTime < it->m_fTime && it->m_fTime <= m_fCurrentTime))
		{
			if (it->m_strID == "Project")
			{
				vector<float> vecDirection;
				vector<string> vecString;
				Util::Parse(m_strStringParams, ',', &vecString);
				for (vector<string>::iterator it = vecString.begin(); it != vecString.end(); ++it)
				{
					vecDirection.push_back(atof((*it).c_str()));
				}

				for (vector<float>::iterator itDirection = vecDirection.begin(); itDirection != vecDirection.end(); ++itDirection)
				{
					int nEntityID = 0;
					int nProjectileMasterDataID = atoi(it->m_vecParams[0].c_str());
					Projectile* pProjectile = NULL;
					m_pGameRoom->CreateProjectile(nProjectileMasterDataID, &nEntityID, &pProjectile, m_pEntity->GetID());
					pProjectile->SetPosition(m_pEntity->GetPosition());

					GameEvent::EntityCreate* pEntityCreate = new GameEvent::EntityCreate();
					pEntityCreate->m_fEventTime = lUpdateTime / 1000.0f;
					pEntityCreate->m_nEntityID = nEntityID;
					pEntityCreate->m_nMasterDataID = nProjectileMasterDataID;
					pEntityCreate->m_EntityType = FBS::Data::EntityType_Projectile;
					pEntityCreate->m_vec3Position = pProjectile->GetPosition();
					pEntityCreate->m_vec3Rotation = pProjectile->GetRotation();

					m_pGameRoom->AddGameEvent(pEntityCreate);

					if (m_pEntity->GetEntityType() == FBS::Data::EntityType::EntityType_Character)
					{
						Character* pCharacter = (Character*)m_pEntity;
						if (pCharacter->GetRole() == Character::Role::Disturber)
						{
							IState* pState = Factory::Instance()->CreateState(m_pGameRoom, pProjectile, StateID::ChallengerDisturbing, lUpdateTime);
							pState->Initialize();
							pProjectile->AddState(pState, lUpdateTime);
							pState->Update(lUpdateTime);
						}
					}

					float fTargetAngle = 0;
					if (m_pTarget == NULL)
					{
						fTargetAngle = m_pEntity->GetRotation().y();
					}
					else
					{
						fTargetAngle = Util::GetAngle_Y(m_pTarget->GetPosition() - m_pEntity->GetPosition());
					}

					btVector3 vec3Dest = Util::GetAngledPosition(m_pEntity->GetPosition(), fTargetAngle + *itDirection, 20/*temp..*/);
					pProjectile->GetBehavior(BehaviorID::MOVE)->Start(lUpdateTime, &vec3Dest);
					pProjectile->GetBehavior(BehaviorID::MOVE)->Update(lUpdateTime);
				}
			}
			else if (it->m_strID == "Fire")
			{
				btVector3 vec3Offset(atof(it->m_vecParams[0].c_str()), atof(it->m_vecParams[1].c_str()), atof(it->m_vecParams[2].c_str()));
				btVector3 vec3Range(atof(it->m_vecParams[3].c_str()), atof(it->m_vecParams[4].c_str()), atof(it->m_vecParams[5].c_str()));
				int nTypes = m_pEntity->GetAttackTargetTypes();
				list<CollisionObject*> objects;
				btVector3 vec3Position = Util::GetAngledPosition(m_pEntity->GetPosition() + vec3Offset, m_pEntity->GetRotation().y(), vec3Range.z() * 0.5f);

				if (m_pGameRoom->GetCollisionObjectsInRange(vec3Position, m_pEntity->GetRotation(), vec3Range, nTypes, &objects))
				{
					for (list<CollisionObject*>::iterator iterCollisionObject = objects.begin(); iterCollisionObject != objects.end(); ++iterCollisionObject)
					{
						CollisionObject* pCollisionObject = *iterCollisionObject;
						int nTargetEntityID = m_pGameRoom->GetEntityIDByCollisionObjectID(pCollisionObject->GetID());

						if (nTargetEntityID == m_pEntity->GetID())
							continue;

						m_pGameRoom->EntityAttack(m_pEntity->GetID(), nTargetEntityID, 1, lUpdateTime);
					}
				}
			}
		}
	}

	if (m_fLength != -1 && m_fCurrentTime >= m_fLength)
	{
		Stop(lUpdateTime - (m_fCurrentTime - m_fLength) * 1000);
	}
}

File: Client_Root/Client/Assets/Scripts/Behavior/Behaviors/Fire.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

namespace Behavior
{
    public class Fire : IBehavior
    {
        IEntity m_Target;

        public override void Initialize(IEntity entity, int nMasterDataID, float fTickInterval)
        {
            m_Entity = entity;
            m_nMasterDataID = nMasterDataID;
            m_fTickInterval = fTickInterval;

            MasterData.Behavior masterBehavior = null;
            MasterDataManager.Instance.GetData<MasterData.Behavior>(nMasterDataID, ref masterBehavior);

            m_fLength = masterBehavior.m_fLength;
            m_strStringParams = masterBehavior.m_strStringParams;
            m_listAction = masterBehavior.m_listAction;
        }

        protected override void UpdateBody(int nUpdateTick)
        {
            foreach(MasterData.Behavior.Action action in m_listAction)
            {
                int nTick = (int)(action.m_fTime / m_fTickInterval);

                if (nUpdateTick == nTick)
                {
                    if (action.m_strID == "Project")
                    {
                        List<float> listDirection = new List<float>();
                        Util.Parse(m_strStringParams, ',', listDirection);

                        foreach(float dir in listDirection)
                        {
                            int nEntityID = 0;
                            int nProjectileMasterDataID = int.Parse(action.m_listParams[0]);
                            Projectile projectile = null;

                            BaeGameRoom2.Instance.CreateProjectile(nProjectileMasterDataID, ref nEntityID, ref projectile, m_Entity.GetID());
                            projectile.SetPosition(m_Entity.GetPosition());

                            if (m_Entity.GetEntityType() == FBS.Data.EntityType.Character)
                            {
                                Character character = (Character)m_Entity;
                                if (character.GetRole() == Character.Role.Disturber)
                                {
                                    IState state = Factory.Instance.CreateState(4);
                                    state.Initialize(projectile, 4, BaeGameRoom2.Instance.GetTickInterval());

                                    projectile.AddState(state, nUpdateTick);

                                    state.StartTick(nUpdateTick);
                                    state.UpdateTick(nUpdateTick);
                                }
                            }

                            float fTargetAngle = 0;
                            if (m_Target == null)
                            {
                                fTargetAngle = m_Entity.GetRotation().y;
                            }
                            else
                            {
                                fTargetAngle = 0;   //  dummy
                            }

                            projectile.GetBehavior(MasterDataDefine.BehaviorID.MOVE).StartTick(nUpdateTick);
                            projectile.GetBehavior(MasterDataDefine.BehaviorID.MOVE).UpdateTick(nUpdateTick);
                        }
                    }
                    else if (action.m_strID == "Fire")
                    {
                    }
                }
            }
        }
    }
}